### PR TITLE
Update the `ICommandPredictor` to provide more feedback and also make feedback easier to be corelated

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Update the `ICommandPredictor` interface APIs to allow a predictor to get feedback about what suggestions from it were displayed to the user, and also make it easier for a predictor to corelate feedback events.

1. Add `ICommandPredictor.OnSuggestionDisplayed`, to let a predictor know what results were displayed to the user when rendering the results: `void OnSuggestionDisplayed(uint session, int countOrIndex);`
    - The `countOrIndex` parameter can help a predictor to tell if the display was in `ListView` or `InlineView`:
       - When the value is `> 0`, it's the number of displayed suggestions from the list returned in <see cref="session"/>, starting from the index 0. This indicates the `ListView` is in use.
       - When the value is `<= 0`, it means a single suggestion from the list got displayed, and the index is the absolute value. This indicates the `InlineView` is in use.

2. Add `clientId` and `session` parameters to the existing interface APIs. The `clientId` parameter helps a predictor to know what host client is calling it (`PSReadLine` only today, but it could be the PS VSCode extension some day in future); the `session` parameter helps a predictor to corelate feedback events.
    - `GetSuggestion` and `StartEarlyProcessing` both take a `string` parameter `clientId`, so it knows the client that makes the call
    - `GetSuggestion` returns a `uint` type `session` value, along with the list of suggestion entries it provides.
    - `OnSuggestionDisplayed` and `OnSuggestionAccepted` both take a `uint` parameter `session`, which indicates where the displayed or accepted suggestions were from.

/cc @kceiw

This PR also includes a few changes to the scripts that generate PowerShell SDK NuGet packages, to make it works properly with the other changes from this PR, and also make NuGet packages include the XML document files.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [x] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): `PSSubsystemPluginModel`
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: No need to update PowerShell docs. It affect the XML documentation only.
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs [hunk 1]
```diff
@@ -40,22 +40,35 @@ public interface ICommandPredictor : ISubsystem
...
+        /// When the value is less than or equal to 0, it means a single suggestion from the list got displayed, and the index is the absolute value.
+        /// </param>
+        void OnSuggestionDisplayed(uint session, int countOrIndex);
+
+        /// <summary>
+        /// The suggestion provided by the predictor was accepted.
+        /// </summary>
+        /// <param name="session">Represents the mini-session where the accepted suggestion came from.</param>
+        /// <param name="acceptedSuggestion">The accepted suggestion text.</param>
+        void OnSuggestionAccepted(uint session, string acceptedSuggestion);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs:71]: We could merge both methods if they are always used together.

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs [hunk 2]
```diff
@@ -27,15 +27,21 @@ public sealed class PredictionResult
         /// </summary>
         public string Name { get; }
 
+        /// <summary>
+        /// Gets the mini-session id that represents a specific invocation to the <see cref="ICommandPredictor.GetSuggestion"/> API of the predictor.
+        /// </summary>
+        public uint Session { get; }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs:33]: Why uint? Maybe Guid to be globally independent?

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs:71]: They are not used together. `OnSuggestionDisplayed` will be fired way more than `OnSuggestionAccepted`.

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs:33]: The session is always under a context, for example, a period of time when a predictor module is loaded and used, a Runspace instance id where the module is loaded. That context can be uniquely identified, so I think there is no need for the mini-session to be globally unique.

I use `uint` only because it provides a large space and easy to increment to get a new mini-session id. Azure PowerShell team is using this prediction API for the [Az Predictor](https://techcommunity.microsoft.com/t5/azure-tools/announcing-az-predictor/ba-p/1873104). I'm waiting on the their feedback on this.

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs:71]: My thoughts are if we get 100 suggestions we could weight all 100 (including displayed and accepted ones) and return the result to the prediction service so that it could make more better analyze. So users could prefer a case-sensitive selection or use other sources for selection (local configs, history or other prediction services) to inform the prediction service.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs:71]: > My thoughts are if we get 100 suggestions we could weight all 100 (including displayed and accepted ones) and return the result to the prediction service so that it could make more better analyze.

It's pretty common that we get suggestions from a predictor but user doesn't accept anything, so it's not clear to me how to always include "displayed and accepted ones" in a feedback.

After this change, "OnSuggestionDisplayed" event will be fired very often. I very much want to reduce it ..., so any ideas are very welcome. Can you maybe elaborate your thoughts more?

> So users could prefer a case-sensitive selection or use other sources for selection (local configs, history or other prediction services) to inform the prediction service.

Not sure I understand what you mean, it would be nice if you can clarify more. Basically, we try to avoid sending more user local infor to a predictor than really needed.

Comment (iSazonov) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs:71]: We've heard more than once that users want to restore a PowerShell session after a restart. For this service it might be useful too.

My thoughts may be vague, but I believe that if a service does not store permanently the (session) state, then this state should be stored locally.

What is a session data? (1) Personal sensitive user data (stored only locally), (2) service data. Service data may be (1) complex data structure (like subset of ML model), (2) simple tips (references on internal service models)

I think that if user data should not be stored on the service, then there might be the following options:
1. No data is stored locally between sessions
2. All session data is stored locally (including model)
2. All sensitive data is stored locally (including service tips). 

Also I think there are many claims about sensing user data to Cloud. The service could be data agnostic and gets encrypted data.

**Comment by daxian-dbw** [comment]
@rjmholt and @anmenaga this PR is ready, please help review it, thank!
@kceiw I will appreciate it if you can review this PR as well.

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/CommandPrediction.cs]: The Az.Predictor team has confirmed that `uint` works fine, becuase they already have other context information that is uniquely identifiable. /cc @kceiw

File: src/System.Management.Automation/engine/Utils.cs [hunk 3]
```diff
@@ -2312,6 +2312,14 @@ internal static void NotNullOrEmpty(string value, string paramName)
             }
         }
 
+        internal static void NotNullOrEmpty(ICollection value, string paramName)
```
Comment (powercode) [inline src/System.Management.Automation/engine/Utils.cs:2315]: Perhaps use `[CallerArgumentExpression("value")]` ?

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/Utils.cs:2315]: It's not implemented yet, right? See the [proposal doc](https://github.com/dotnet/csharplang/blob/master/proposals/caller-argument-expression.md) and the open issue: https://github.com/dotnet/csharplang/issues/287

File: src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs (hunk 1 above)
Comment (anmenaga) [reply src/System.Management.Automation/engine/Subsystem/CommandPrediction/ICommandPredictor.cs]: This feels like a design proposal for a separate feature "restore a PowerShell session after a restart" :)

File: tools/packaging/packaging.psm1 [hunk 4]
```diff
@@ -2268,6 +2276,9 @@ function New-ReferenceAssembly
             Copy-Item $refBinPath $RefAssemblyDestinationPath -Force
             Write-Log "Reference assembly '$assemblyName.dll' built and copied to $RefAssemblyDestinationPath"
 
+            Copy-Item $dllXmlDoc $RefAssemblyDestinationPath -Force
+            Write-Log "Xml document '$assemblyName.xml' copied to $RefAssemblyDestinationPath"
```
Comment (anmenaga) [inline tools/packaging/packaging.psm1:2280]: `... make NuGet packages include the XML document files.`
This is not really related to prediction APIs. Could have been a separate PR.

File: tools/packaging/packaging.psm1 [hunk 5]
```diff
@@ -2367,6 +2378,11 @@ function CleanupGeneratedSourceCode
             Pattern = "[System.Runtime.CompilerServices.CompilerGeneratedAttribute, System.Runtime.CompilerServices.NullableContextAttribute((byte)2)]"
             Replacement = "/* [System.Runtime.CompilerServices.CompilerGeneratedAttribute, System.Runtime.CompilerServices.NullableContextAttribute((byte)2)] */ "
         },
+        @{
+            ApplyTo = @("System.Management.Automation")
+            Pattern = "[System.Runtime.CompilerServices.CompilerGeneratedAttribute, System.Runtime.CompilerServices.IsReadOnlyAttribute]"
+            Replacement = "/* [System.Runtime.CompilerServices.CompilerGeneratedAttribute, System.Runtime.CompilerServices.IsReadOnlyAttribute] */ "
```
Comment (anmenaga) [inline tools/packaging/packaging.psm1:2384]: Can you please explain why this is needed?

**Review by anmenaga — APPROVED** [review]
Looks fine.
Please file a doc issue and link it in "Documentation needed"/"Issue filed:" in PR Checklist; thank you.

File: src/System.Management.Automation/engine/Utils.cs (hunk 3 above)
Comment (powercode) [reply src/System.Management.Automation/engine/Utils.cs:2315]: You're correct! I was misled by the docs.

File: tools/packaging/packaging.psm1 (hunk 4 above)
Comment (daxian-dbw) [reply tools/packaging/packaging.psm1:2280]: Not directly related, but sort of. The script needs to be updated after the changes in API because new attributes are used which need to be scrubbed when generating the reference assembly, so I have to update the packaging scripts. While doing so, I found the XML docs are missing when using our NuGet packages in VS Code or Visual Studio, and thus I fix that too.
Since we need to include part of the fixes in `packaging.psm1` anyways, I prefer to keep all changes in this PR.

File: tools/packaging/packaging.psm1 (hunk 5 above)
Comment (daxian-dbw) [reply tools/packaging/packaging.psm1:2384]: The `CompilerGeneratedAttribute` and `IsReadOnlyAttribute` are added by the C# compiler to the readonly members of a struct in the generated IL. Those attributes can only be used by the C# compiler and thus will result in complication error when they are in the source code. Therefore, we need to scrub them out before compiling to get our reference assembly.

**Comment by daxian-dbw** [comment]
@anmenaga Thanks for the review!
There is no need to update PowerShell docs. The `PSSubsystemPluginModel` experimental feature is only mentioned here: https://docs.microsoft.com/en-us/powershell/scripting/learn/experimental-features?view=powershell-7.1#pssubsystempluginmodel, and it doesn't talk about specific APIs. We only need to update the XML document for the APIs, which is done by the updates to `packaging.psm1`.
